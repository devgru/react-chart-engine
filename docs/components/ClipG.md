# ClipG

Компонент используется для ограничения «рабочей» зоны графика в которой происходит отрисовка содержимого визуалиации. Исключает возможность графических элементов выходить за границы графика (например, «залезать» на оси).

Компонент рендерит `g`-элемент. Потомки, переданные в элемент `ClipG` будут помещены во вложенный `g`, с зоной видимости ограниченной прямоугольным `clip-path`, с размерами `width` и `height`.

Также создаёт элемент `clip-path` с `id` равным переданному свойству `id` с добавлением строки `.clip`. Этот `clip-path` можно использовать так же для других векторных элементов на странице.

## Свойства

Все свойства **обязательны**.

- строка `id`, уникальный в рамках страницы идентификатор;
- число `width`, ширина «рабочей» зоны графика;
- число `height`, высота «рабочей» зоны графика.
