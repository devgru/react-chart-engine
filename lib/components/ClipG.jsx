import React from 'react';
import PropTypes from 'prop-types';

function ClipG({ id, width, height, children }) {
  return (
    <g>
      <defs>
        <clipPath id={`${id}.clip`}>
          <rect width={width} height={height} />
        </clipPath>
      </defs>
      <g clipPath={`url(#${id}.clip)`}>
        {children}
      </g>
    </g>
  );
}

ClipG.propTypes = {
  children: PropTypes.node,

  id: PropTypes.string.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
};

ClipG.defaultProps = {
  children: null,
};

export default ClipG;
