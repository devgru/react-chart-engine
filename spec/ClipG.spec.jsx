import ClipG from '../lib/components/ClipG';

import React from 'react';
import renderer from 'react-test-renderer';

describe('ClipG', () => {
  it('should work with props', () => {
    const component = renderer.create(
      <ClipG id='clip' width={100} height={100}>
        <circle/>
      </ClipG>);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
