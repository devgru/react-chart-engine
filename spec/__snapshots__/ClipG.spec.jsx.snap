exports[`ClipG should work with props 1`] = `
<g>
  <defs>
    <clipPath
      id="clip.clip">
      <rect
        height={100}
        width={100} />
    </clipPath>
  </defs>
  <g
    clipPath="url(#clip.clip)">
    <circle />
  </g>
</g>
`;
